### PR TITLE
Jenkins Pipelines Lab

### PR DESCRIPTION
**Pull request by chrira**
Re-add Jenkins based pipelines lab for OCP 4

### CODE REVIEW

File: labs/13_template_creation.md [hunk 1]
```diff
@@ -11,6 +11,12 @@ Auflisten aller Befehle:
 oc help
 ```
 
+Konzepte und Typen:
+
+```
+$ oc types
+```
```
Comment (bliemli) [inline labs/13_template_creation.md]: `oc types` doesn't seem to exist in my `oc` 4.6.9 client.

File: additional-labs/jenkins.md [hunk 2]
```diff
@@ -0,0 +1,407 @@
+# Jenkins Pipelines auf OpenShift
+
+Mit Jenkins Pipelines auf OpenShift hat man die Möglichkeit, komplexe CI/CD Prozesse voll integriert abzubilden. In diesem Lab zeigen wir, wie man mit Jenkins Pipelines arbeitet und so Applikationen buildet, testet und entsprechend kontrolliert in die verschiedenen Stages deployt.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Mit Jenkins Pipelines auf OpenShift hat man die Möglichkeit, komplexe CI/CD Prozesse voll integriert abzubilden. In diesem Lab zeigen wir, wie man mit Jenkins Pipelines arbeitet und so Applikationen baut, testet und entsprechend kontrolliert in die verschiedenen Stages deployt.
```

File: additional-labs/jenkins.md [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+
+Die offizielle Dokumentation ist unter [Pipeline build](https://docs.openshift.com/container-platform/latest/builds/build-strategies.html#builds-strategy-pipeline-build_build-strategies) oder [OpenShift 3 Pipeline Builds](https://docs.openshift.com/container-platform/3.11/dev_guide/dev_tutorials/openshift_pipeline.html) zu finden.
+
+Die OpenShift Plattform setzt, seit Version 4, für die integrierten Pipeline Builds auf Tekton.
+Diese, Tekton basierten, Pipelines sind jedoch erst als Technology Preview Feature verfügbar [OpenShift 4 Pipeline Builds](https://docs.openshift.com/container-platform/latest/pipelines/understanding-openshift-pipelines.html).
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Diese Tekton-basierten Pipelines sind jedoch erst als Technology Preview Feature verfügbar [OpenShift 4 Pipeline Builds](https://docs.openshift.com/container-platform/latest/pipelines/understanding-openshift-pipelines.html).
```

File: additional-labs/jenkins.md [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+
+## LAB: Eine einfache OpenShift Pipeline anlegen und ausführen
+
+Um zu verstehen, wie OpenShift Pipelines funktionieren, wollen wir als ersten Schritt direkt eine Pipeline anlegen.
+
+Erstellen wir dafür ein neues Projekt mit dem Namen `[USER]-buildpipeline`.
+
+<details><summary>Tipp</summary>oc new-project [USER]-buildpipeline</details><br/>
+
+Jetzt deployen wir Jenkins in user Projekt:
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Jetzt deployen wir Jenkins in unser Projekt:
```

File: additional-labs/jenkins.md [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+Wenn Sie zum ersten Mal einen OpenShift Jenkins aufrufen, erhalten Sie eine OpenShift-Authentifizierungsmaske.
+
+![Jenkins OAuth2 Login](../images/jenkins_oauth2_login.png "Jenkins OAuth2 Login")
+
+Danach wird ein Berechtigungs-Bildschirm angezeigt, in dem Sie nach Autorisierungs-Berechtigungen gefragt werden.
+Siehe das Bild unten.
+
+![Jenkins OAuth2 permissions](../images/jenkins_oauth2_permissions.png "Jenkins OAuth2 permissions")
+
+Akzeptieren Sie diese und wechseln Sie zum nächsten Bildschirm, indem Sie auf 'Alow selected permissions' klicken.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Akzeptieren Sie diese und wechseln Sie zum nächsten Bildschirm, indem Sie auf 'Allow selected permissions' klicken.
```

File: additional-labs/jenkins.md [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+
+Der nächste Bildschirm sollte der berühmte/klassische (oder berüchtigte) Jenkins-Begrüssungsbildschirm sein.
+
+![Jenkins welcome](../images/pipeline-jenkins-overview.png "Jenkins welcome")
+
+Die Pipeline aus der BuildConfig ist im Ordner vom unserem OpenShift Projekt sichtbar.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Die Pipeline aus der BuildConfig ist im Ordner unseres OpenShift Projekts sichtbar.
```

File: additional-labs/jenkins.md [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+        sleep 5
+    }
+    stage('PromoteProd') {
+        echo "Deploy to Prod"
+        sleep 5
+    }
+}
+```
+
+Die Testpipeline besteht aus sechs Pipeline-Stages, welche auf zwei Jenkins Slaves ausgeführt werden. Die `Build` Pipeline-Stage, ist aktuell die einzig ausprogrammierte Stage. Sie startet nämlich im aktuellen OpenShift Projekt den Docker Build für unsere Applikation und wartet entsprechend bis dieser erfolgreich abgeschlossen ist.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Die Testpipeline besteht aus sechs Pipeline-Stages, welche auf zwei Jenkins Slaves ausgeführt werden. Die `Build` Pipeline-Stage ist aktuell die einzig ausprogrammierte Stage. Sie startet nämlich im aktuellen OpenShift Projekt den Image Build für unsere Applikation und wartet entsprechend bis dieser erfolgreich abgeschlossen ist.
```

File: additional-labs/jenkins.md [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+
+<details><summary>Tipp</summary>oc new-project [USER]-buildpipeline</details><br/>
+
+Jetzt deployen wir Jenkins in user Projekt:
+
+```bash
+oc get template/jenkins-ephemeral -n openshift -o yaml | oc process --param MEMORY_LIMIT=2Gi --param DISABLE_ADMINISTRATIVE_MONITORS=true -f - | oc apply -f -
+```
+
+Wir legen mit folgendem Befehl die entsprechende BuildConfig an, welche das JenkinsFile, also die Pipeline, direkt beinhaltet. Ebenso wird eine zweite BuildConfig erstellt. Diese enthält die Docker BuildConfig für die eigentliche Applikation, die wir im Rahmen dieser Pipeline deployen wollen. Im vorliegenden Beispiel eine simple PHP Applikation:
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Wir legen mit folgendem Befehl die entsprechende BuildConfig an, welche das JenkinsFile, also die Pipeline, direkt beinhaltet. Ebenso wird eine zweite BuildConfig erstellt. Diese enthält die BuildConfig für die eigentliche Applikation, die wir im Rahmen dieser Pipeline deployen wollen. Im vorliegenden Beispiel eine simple PHP Applikation:
```

File: additional-labs/jenkins.md [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+
+Will man selber Jenkins Slaves builden, sollte man auf dem `openshift/jenkins-slave-base-centos7` Image basieren.
+
+## LAB: Multi-Stage Deployment
+
+Als nächstes wollen wir nun unsere Pipeline weiter ausbauen und das Deployment der Applikation auf den unterschiedlichen Stages (dev, test, prod) angehen.
+
+Für ein Multi-Stage Deployment auf OpenShift hat sich das folgende Setup als Best Practice erwiesen:
+
+- Ein Build Projekt, CI/CD Umgebung Jenkins, Docker Builds, S2I Builds, ...
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
- Ein Build Projekt, CI/CD Umgebung Jenkins, Image Builds, S2I Builds, ...
```

File: additional-labs/jenkins.md [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+
+Die Testpipeline besteht aus sechs Pipeline-Stages, welche auf zwei Jenkins Slaves ausgeführt werden. Die `Build` Pipeline-Stage, ist aktuell die einzig ausprogrammierte Stage. Sie startet nämlich im aktuellen OpenShift Projekt den Docker Build für unsere Applikation und wartet entsprechend bis dieser erfolgreich abgeschlossen ist.
+
+Die letzten drei Steps werden bedingt durch den Node Selector `node ('maven') { ... }` auf einem Jenkins Slave mit dem Namen `maven` ausgeführt. Dabei startet OpenShift mittels Kubernetes Plugin dynamisch einen Jenkins Slave Pod und führt entsprechend diese Build-Stages auf diesem Slave aus.
+
+Der `maven` Slave ist im Kubernetes Cloud Plugin vom Jenkins vorkonfiguriert. Weiter unten im Kapitel zu Custom Slaves werden Sie lernen, wie man individuelle Slaves dazu verwendet um Pipelines darauf auszuführen.
+
+## BuildConfig Optionen
+
+Im vorherigen Lab haben wir in der BuildConfig vom Typ `jenkinsPipelineStrategy` das JenkinsFile direkt angegeben. Als Alternative dazu kann das Jenkins Pipeline File auch via GitRepository in der BuildConfig hinterlegt werden.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Im vorherigen Lab haben wir in der BuildConfig vom Typ `jenkinsPipelineStrategy` das JenkinsFile direkt angegeben. Als Alternative dazu kann das Jenkins Pipeline File auch via Git Repository in der BuildConfig hinterlegt werden.
```

File: additional-labs/jenkins.md [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+
+## Jenkins OpenShift Plugins
+
+Der durch OpenShift dynamisch deployte Jenkins ist durch eine Reihe von OpenShift Jenkins Plugins vollständig mit OpenShift gekoppelt. Einerseits kann so direkt auf Ressourcen innerhalb des Projekts zugegriffen werden, andererseits können durch entsprechendes Labelling dynamische Slaves aufgesetzt werden. Des Weiteren wird auch ein entsprechender Serviceaccount (`jenkins`) erstellt. Die Rechtevergabe kann entsprechend über diesen Serviceacount erfolgen.
+
+Zusätzliche Informationen finden Sie hier: <https://docs.openshift.com/container-platform/latest/builds/build-strategies.html#builds-strategy-pipeline-build_build-strategies>
+
+### OpenShift Jenkins Pipeline
+
+Mit dem OpenShift Jenkins Client Plugin kann so auf einfache Art direkt mit dem OpenShift Cluster kommuniziert werden und als Jenkinsfile komplexe CI/CD Pipelines implementieren:
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Mit dem OpenShift Jenkins Client Plugin kann so auf einfache Art direkt mit dem OpenShift Cluster kommuniziert und als Jenkinsfile komplexe CI/CD Pipelines implementiert werden:
```

File: additional-labs/jenkins.md [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+
+```bash
+oc import-image custom-jenkins-slave --from=docker.io/openshift/jenkins-slave-maven-centos7 --confirm
+```
+
+Mit diesem Befehl haben wir einen ImageStream zu diesem Image angelegt. Wir inspizieren ihn über die Web Console oder mit dem oc Tool.
+
+<details><summary>Tipp</summary>Administrator Web Console: Builds -> Image Streams -> custom-jenkins-slave<br/>oc Tool: oc describe is/custom-jenkins-slave</details><br/>
+
+Anschliessend definieren wir mittels dem label `role` noch, dass es sich um einen Jenkins Slave handelt:
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Anschliessend definieren wir mittels dem Label `role` noch, dass es sich um einen Jenkins Slave handelt:
```

File: additional-labs/jenkins.md [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+    stage('custom stage') {
+        echo "Running on the custom jenkins slave"
+    }
+}
+```
+
+Erweitern Sie ihre Pipeline mit dieser Stage entweder im Jenkins Job, der Web Console oder in der BuildConfig.
+
+__Note__:
+Der einfachste Weg ist die Pipeline im Jenkins GUI zu Erweitern.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Der einfachste Weg ist die Pipeline im Jenkins GUI zu erweitern.
```

File: additional-labs/jenkins.md [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+Dazu den Job `appuio-sample-pipeline` auswählen und `Configure` anklicken.
+Dort ist der Code der Pipeline in einem Editor Fenster veränderbar.
+Einfach das `custom-jenkins-slave` Code Snippet am Schluss der Pipeline einfügen.
+Mit `Build Now` einen neuen Build starten.
+
+Überprüfen Sie, dass die neue Stage ausgeführt wurde.
+
+Über diesen Mechanismus lassen sich beliebige - auch selbst gebuildete - Jenkins Slaves aufsetzen und verwenden. Es stehen out-of-the-box ein `maven und gradle` sowie ein `NodeJS` Slave zur Verfügung.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Über diesen Mechanismus lassen sich beliebige - auch selbst gebaute - Jenkins Slaves aufsetzen und verwenden. Es stehen out-of-the-box ein `maven` und `gradle` sowie ein `NodeJS` Slave zur Verfügung.
```

File: additional-labs/jenkins.md [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+
+Will man selber Jenkins Slaves builden, sollte man auf dem `openshift/jenkins-slave-base-centos7` Image basieren.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Will man selber Jenkins Slaves bauen, sollte man auf dem `openshift/jenkins-slave-base-centos7` Image basieren.
```

Comment (bliemli) [reply additional-labs/jenkins.md]: Ist centos7 wirklich noch ratsam?

File: additional-labs/jenkins.md [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+
+## LAB: Multi-Stage Deployment
+
+Als nächstes wollen wir nun unsere Pipeline weiter ausbauen und das Deployment der Applikation auf den unterschiedlichen Stages (dev, test, prod) angehen.
+
+Für ein Multi-Stage Deployment auf OpenShift hat sich das folgende Setup als Best Practice erwiesen:
+
+- Ein Build Projekt, CI/CD Umgebung Jenkins, Docker Builds, S2I Builds, ...
+- Pro Stage (dev, test, ..., prod) ein Projekt, welches die laufenden Pods und Services enthält, was uns erlaubt, sämtliche Umgebungen möglichst identisch zu halten
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
- Pro Stage (dev, test, ..., prod) ein Projekt, welches die laufenden Pods und Services enthält, was uns erlaubt, sämtliche Umgebungen praktisch identisch zu halten
```

File: additional-labs/jenkins.md [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+- [USER]-pipeline-prod
+
+<details>
+    <summary>Tipp</summary>
+    oc new-project [USER]-pipeline-dev<br/>
+    oc new-project [USER]-pipeline-test<br/>
+    oc new-project [USER]-pipeline-prod
+</details><br/>
+
+Nun müssen wir den `puller` Serviceaccount aus den entsprechenden Projekten pro Stage die nötigen Rechte geben, damit die gebuildeten Images gepullt werden können.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Nun müssen wir dem `puller` Serviceaccount aus den entsprechenden Projekten pro Stage die nötigen Rechte geben, damit die gebuildeten Images gepullt werden können.
```

File: additional-labs/jenkins.md [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+
+Als nächstes erstellen wir in den Stage-Projekten die Applikationen. Dafür definieren wir einen Tag im ImageStream, welcher deployt werden soll. Zuvor müssen wir aber die entsprechenden Tags erstellen, die deployt werden sollen:
+
+```bash
+oc tag [USER]-buildpipeline/application:latest [USER]-buildpipeline/application:dev
+oc tag [USER]-buildpipeline/application:dev [USER]-buildpipeline/application:test
+oc tag [USER]-buildpipeline/application:test [USER]-buildpipeline/application:prod
+```
+
+Erstellen der Applikationen.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
### Erstellen der Applikationen
```

File: additional-labs/jenkins.md [hunk 19]
```diff
@@ -0,0 +1,407 @@
...
+
+In jeder Stage erstellen wir die Applikation anhand des vorher angelegten ImageStreams.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
In jeder Stage erstellen wir nun die Applikation anhand des vorher angelegten ImageStreams.
```

File: additional-labs/jenkins.md [hunk 20]
```diff
@@ -0,0 +1,407 @@
...
+Achtung, die Applikationen müssen im richtigen Projekt erstellt werden.
+
+<details>
+    <summary>Tipp</summary>
+    dev: oc new-app [USER]-buildpipeline/application:dev -n [USER]-pipeline-dev<br/>
+    test: oc new-app [USER]-buildpipeline/application:test -n [USER]-pipeline-test<br/>
+    prod: oc new-app [USER]-buildpipeline/application:prod -n [USER]-pipeline-prod
+</details><br/>
+
+Danach können wir noch die Applikation exposen.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Danach können wir noch die Applikation exponieren.
```

File: additional-labs/jenkins.md [hunk 21]
```diff
@@ -0,0 +1,407 @@
...
+
+<details>
+    <summary>Tipp</summary>
+    dev: oc create route edge --service=application -n [USER]-pipeline-dev<br/>
+    test: oc create route edge --service=application -n [USER]-pipeline-test<br/>
+    prod: oc create route edge --service=application -n [USER]-pipeline-prod
+</details><br/>
+
+In der Pipeline können wir nun mittels Setzen eines bestimmten Tags auf dem Imagestream der gebuildeten Applikation bspw. `application:dev`, das entsprechende Image in die passende Stage promoten und deployen.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
In der Pipeline können wir nun mittels Setzen eines bestimmten Tags auf dem Imagestream der gebauten Applikation, bspw. `application:dev`, das entsprechende Image in die passende Stage promoten und deployen.
```

File: additional-labs/jenkins.md [hunk 22]
```diff
@@ -0,0 +1,407 @@
...
+                // trigger Deployment in prod project
+                def dc = openshift.selector('deploy', "application")
+                dc.rollout().status()
+            }
+        }
+    }
+}
+```
+
+Führen Sie die Pipeline erneut aus und schauen Sie sich an, wie nun die gebuildete Applikation von Stage zu Stage deployt wird.
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Führen Sie die Pipeline erneut aus und schauen Sie sich an, wie nun die gebaute Applikation von Stage zu Stage deployt wird.
```

File: additional-labs/jenkins.md [hunk 23]
```diff
@@ -0,0 +1,407 @@
...
+
+![Run Pipeline Jenkins](../images/openshift-pipeline-rollout.png)
+
+## Jenkins Pipeline Sprache
+
+Unter <https://github.com/puzzle/jenkins-techlab> finden Sie ein entsprechendes Hands-on Lab zur Jenkins Pipeline Sprache. Die Syntax ist [hier](https://jenkins.io/doc/book/pipeline/syntax/) beschrieben.
+
+## Deployment von Resourcen und Konfiguration
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
## Deployment von Ressourcen und Konfiguration
```

File: additional-labs/jenkins.md [hunk 24]
```diff
@@ -0,0 +1,407 @@
...
+
+Bisher haben wir nur die Applikation mittels unserer Deployment Pipeline deployt. Wie können wir nun unsere Pipeline auch dazu verwenden, um Resourcen (Deployments, Routen, Cronjobs...) zu deployen?
```
Comment (bliemli) [inline additional-labs/jenkins.md]:
```suggestion
Bisher haben wir nur die Applikation mittels unserer Deployment Pipeline deployt. Wie können wir nun unsere Pipeline auch dazu verwenden, um Ressourcen (Deployments, Routen, Cronjobs...) zu deployen?
```

**Review by bliemli — CHANGES_REQUESTED** [review]

File: labs/13_template_creation.md (hunk 1 above)
Comment (chrira) [reply labs/13_template_creation.md]: thanks for the hint

**Review by bliemli — APPROVED** [review]

**Pull request merged**